### DISCUSSION TITLE
String conversions for custom enums / structs

### DISCUSSION THREAD

**Discussion by mitchgrout** [slint-ui/slint#4737] [Q&A] [answered]
Is it possible to define a conversion to `string` for a custom enum/struct? If not, is there a particular pattern used to provide this type of conversion explicitly? My use case is, I would like to represent a piano, with all of the white keys labelled with their note and octave for accessibility. My `.slint` file is currently:

```js
export enum BaseNote   { a, b, c, d, e, f, g }
export enum Accidental { flat, neutral, sharp }

export struct Note
{
  octave:     int,
  note:       BaseNote,
  accidental: Accidental,
}

export component PianoKey inherits Rectangle{
  in property<Note> note;

  border-color: black;
  border-width: 1px;
  background:   (note.accidental == Accidental.neutral)? white : black;

  if (note.accidental == Accidental.neutral) : Text {
    text: ?;
  }
}
```

Preferably, I could fill this hole with `note`, and and provide my `struct Note`  with a suitable conversion operator into `string`. I haven't found any notes on this in the documentation, and my second option of having a `pure function to-string() -> string` does not appear to work, as it seems `struct` cannot contain `function` declarations.

**Comment by tronical** [comment, marked as answer]
Interesting point. It sounds like not necessarily a conversion to a string as-is but rather a conversion to a string for the purpose of displaying to the user - which fits into the needs for a UI.

That said, I think perhaps it would make more sense to provide this as a component? Like

```
struct Note {
    ...
}

export component NoteLabel {
    in property <Note> note;
    HorizontalLayout {
        Text {
            text: ...;
        }
        Image {
            source: icon-for-octave(note.octave);
        }
    }
}
```

So perhaps that gives you more freedom in the future beyond text, while still providing you with some degree of encapsulation?

**Reply by mitchgrout** [reply]
Yeah, I could see that; keep the stringification logic tied closely to the component that wants to use it for rendering, as it's possible it could change per-component. I could still see a benefit to allowing `struct`'s to carry functions of their own to allow for things like...

```js
export struct User {
  first-name: string,
  last-name: string,
  function full-name() -> string { ... }
}
```

...but the pros and cons of this may be a little out of scope here

**Reply by tronical** [reply]
Yes, true, member functions on structs seem like a good idea.
